File: python_files/raylib/single_file_games/pong.py
import pyray as p
from math import sqrt


SCREEN_WIDTH = 800
SCREEN_HEIGHT = 600
SCREEN_TITLE = "Pong"
SCREEN_BACKGROUND = p.SKYBLUE
GAME_FPS = 60


class Paddle:
    def __init__(self, x) -> None:
        self.width = 10
        self.height = 100
        self.initial_x = x
        self.initial_y = p.get_screen_height() / 2 - self.height / 2
        self.x = self.initial_x
        self.y = self.initial_y
        self.color = p.BLACK
        self.speed = 5
        self.ai_speed = 7

    def reset(self):
        self.x = self.initial_x
        self.y = self.initial_y

    def draw(self):
        p.draw_rectangle_rec((self.x, self.y, self.width, self.height), self.color)

    def update(self):
        if p.is_key_down(p.KeyboardKey.KEY_W) and self.y >= 0:
            self.y -= self.speed
        elif (
            p.is_key_down(p.KeyboardKey.KEY_S)
            and self.y <= p.get_screen_height() - self.height
        ):
            self.y += self.speed

    def update_ai(self, ball):
        if ball.x > p.get_screen_width() / 2:
            distance_x = ball.x - (self.x + self.width / 2)
            distance_y = ball.y - (self.y + self.height / 2)
            distance = sqrt((distance_x**2) + (distance_y**2))
            if distance > 0:
                distance_y /= distance

                self.y += distance_y * self.ai_speed

        if self.y <= 0:
            self.y = 0
        elif self.y >= p.get_screen_height() - self.height:
            self.y = p.get_screen_height() - self.height


class Ball:
    def __init__(self) -> None:
        self.initial_x = p.get_screen_width() / 2
        self.initial_y = p.get_screen_height() / 2
        self.x = self.initial_x
        self.y = self.initial_y
        self.radius = 10
        self.color = p.RED
        self.initial_speed_x = 4.0
        self.initial_speed_y = 4.0
        self.speed_x = self.initial_speed_x
        self.speed_y = self.initial_speed_y
        self.speed_increment = 0.0005
        self.active = False

    def reset(self):
        self.x = self.initial_x
        self.y = self.initial_y
        self.speed_x = self.initial_speed_x
        self.speed_y = self.initial_speed_y
        self.speed_x *= -1
        self.active = False

    def draw(self):
        p.draw_circle_v((self.x, self.y), self.radius, self.color)

    def update(self):
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.active = True

        if self.active:
            self.x += self.speed_x
            self.y += self.speed_y

            # increment speed
            if self.speed_x > 0:
                self.speed_x += self.speed_increment
            else:
                self.speed_x -= self.speed_increment

            if self.speed_y > 0:
                self.speed_y += self.speed_increment
            else:
                self.speed_y -= self.speed_increment

        if self.y <= self.radius or self.y >= p.get_screen_height() - self.radius:
            self.speed_y *= -1


class Game:
    def __init__(self) -> None:
        self.score_right = 0
        self.score_left = 0

        self.paddle_left = Paddle(10)
        self.ai = Paddle(p.get_screen_width() - self.paddle_left.width - 10)
        self.ball = Ball()

    def reset(self):
        self.ball.reset()
        self.paddle_left.reset()
        self.ai.reset()

    def draw(self):
        # draw scores
        p.draw_text(str(self.score_left), 200, 20, 40, p.BLACK)
        p.draw_text(str(self.score_right), p.get_screen_width() - 220, 20, 40, p.BLACK)

        self.paddle_left.draw()
        self.ai.draw()
        self.ball.draw()

    def update(self):
        self.paddle_left.update()
        self.ai.update_ai(self.ball)
        self.ball.update()

        # update scores
        if self.ball.x < 0:
            self.score_right += 1
            self.reset()
        elif self.ball.x > p.get_screen_width():
            self.score_left += 1
            self.reset()

        # ball collision paddle
        if p.check_collision_circle_rec(
            (self.ball.x, self.ball.y),
            self.ball.radius,
            (
                self.paddle_left.x,
                self.paddle_left.y,
                self.paddle_left.width,
                self.paddle_left.height,
            ),
        ):
            self.ball.speed_x *= -1
        elif p.check_collision_circle_rec(
            (self.ball.x, self.ball.y),
            self.ball.radius,
            (
                self.ai.x,
                self.ai.y,
                self.ai.width,
                self.ai.height,
            ),
        ):
            self.ball.speed_x *= -1


def main():
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)
    p.set_target_fps(GAME_FPS)

    game = Game()

    while not p.window_should_close():
        game.update()

        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()

File: cpp_files/raylib/single_file_games/pong.cpp
#include "raylib.h"
#include <cmath>
#include <string>

// **********************************
// BALL
// **********************************

class Ball
{
public:
    float initial_x;
    float initial_y;
    float x;
    float y;
    float radius;
    Color color;
    float initial_speed_x;
    float initial_speed_y;
    float speed_x;
    float speed_y;
    float speed_increment;
    bool active;

    Ball()
    {
        initial_x = static_cast<float>(GetScreenWidth()) / 2;
        initial_y = static_cast<float>(GetScreenHeight()) / 2;
        x = initial_x;
        y = initial_y;
        radius = 10.0f;
        color = RED;
        initial_speed_x = 4.0f;
        initial_speed_y = 4.0f;
        speed_x = initial_speed_x;
        speed_y = initial_speed_y;
        speed_increment = 0.005f;
        active = false;
    }

    void reset()
    {
        x = initial_x;
        y = initial_y;
        speed_x = initial_speed_x;
        speed_y = initial_speed_y;
        speed_x *= -1;
        active = false;
    }

    void draw() const
    {
        DrawCircleV(Vector2{x, y}, radius, color);
    }

    void update()
    {
        if (IsKeyPressed(KEY_SPACE))
        {
            active = true;
        }

        if (active)
        {
            x += speed_x;
            y += speed_y;

            // increment speed
            if (speed_x > 0)
            {
                speed_x += speed_increment;
            }
            else if (speed_x < 0)
            {
                speed_x -= speed_increment;
            }
            else if (speed_y > 0)
            {
                speed_y += speed_increment;
            }
            else if (speed_y > 0)
            {
                speed_y -= speed_increment;
            }
        }

        if (y <= radius || y >= GetScreenHeight() - radius)
        {
            speed_y *= -1;
        }
    }
};

// **********************************
// PADDLE
// **********************************

class Paddle
{
public:
    float initial_x;
    float initial_y;
    float x;
    float y;
    float width;
    float height;
    Color color;
    float speed;
    float ai_speed;

    Paddle(float x)
    {
        width = 10.0f;
        height = 100.0f;
        initial_x = x;
        initial_y = static_cast<float>(GetScreenHeight()) / 2 - height / 2;
        this->x = initial_x;
        this->y = initial_y;
        color = BLACK;
        speed = 5.0f;
        ai_speed = 7.0f;
    }

    void reset()
    {
        x = initial_x;
        y = initial_y;
    }

    void draw() const
    {
        DrawRectangleRec(Rectangle{x, y, width, height}, color);
    }

    void update_left()
    {
        if (IsKeyDown(KEY_W) && y >= 0)
        {
            y -= speed;
        }
        else if (IsKeyDown(KEY_S) && y <= GetScreenHeight() - height)
        {
            y += speed;
        }
    }

    void update_ai(Ball &ball)
    {
        if (ball.x > GetScreenWidth() / 2)
        {
            float distance_x = ball.x - (x + width / 2);
            float distance_y = ball.y - (y + height / 2);
            float distance = std::sqrt(std::pow(distance_x, 2) + std::pow(distance_y, 2));
            if (distance > 0)
            {
                distance_y /= distance;

                y += distance_y * ai_speed;
            }

            if (y <= 0)
            {
                y = 0;
            }
            else if (y >= GetScreenHeight() - height)
            {
                y = GetScreenHeight() - height;
            }
        }
    }
};

// **********************************
// GAME
// **********************************

class Game
{
public:
    int score_left;
    int score_right;

    Paddle paddle_left;
    Paddle paddle_ai;
    Ball ball;

    Game() : paddle_left(10),
             paddle_ai(GetScreenWidth() - paddle_left.width - 10),
             ball()
    {
        score_left = 0;
        score_right = 0;
    }

    void draw() const
    {
        // draw scores
        DrawText((std::to_string(score_left).c_str()), 200, 30, 40, BLACK);
        DrawText((std::to_string(score_right).c_str()), GetScreenWidth() - 220, 30, 40, BLACK);

        paddle_left.draw();
        paddle_ai.draw();
        ball.draw();
    }

    void update()
    {
        paddle_left.update_left();
        paddle_ai.update_ai(ball);
        ball.update();

        // ball collision paddle
        if (CheckCollisionCircleRec(Vector2{ball.x, ball.y},
                                    ball.radius,
                                    Rectangle{paddle_left.x,
                                              paddle_left.y,
                                              paddle_left.width,
                                              paddle_left.height}))
        {
            ball.speed_x *= -1;
        }
        else if (CheckCollisionCircleRec(Vector2{ball.x, ball.y},
                                         ball.radius,
                                         Rectangle{paddle_ai.x,
                                                   paddle_ai.y,
                                                   paddle_ai.width,
                                                   paddle_ai.height}))
        {
            ball.speed_x *= -1;
        }

        // update scores
        if (ball.x < 0)
        {
            score_right++;
            ball.reset();
            paddle_left.reset();
            paddle_ai.reset();
        }
        else if (ball.x > GetScreenWidth())
        {
            score_left++;
            ball.reset();
            paddle_left.reset();
            paddle_ai.reset();
        }
    }
};

// **********************************
// MAIN
// **********************************

int main(void)
{
    const int screenWidth = 800;
    const int screenHeight = 600;
    const char screenTitle[] = "Pong";
    const Color screenBackground = SKYBLUE;
    const int gameFps = 60;

    InitWindow(screenWidth, screenHeight, screenTitle);
    SetTargetFPS(gameFps);

    Game game;

    while (!WindowShouldClose())
    {
        BeginDrawing();
        ClearBackground(screenBackground);

        game.draw();
        game.update();

        EndDrawing();
    }

    CloseWindow();

    return 0;
}
